File: src/instrument/channel_manager/ChannelManager.cpp
#include "ChannelManager.h"
#include "storage/DatabaseManager.h"
#include <QDebug>

ChannelManager::ChannelManager(QObject* parent)
    : QObject(parent)
{
}

void ChannelManager::loadChannels()
{
    m_channels.clear();

    auto channels = DatabaseManager::instance().localDb()->loadAllChannels();
    for (auto& ch : channels) {
        m_channels.insert(ch.channelId, ch);
    }

    qDebug() << "Loaded" << m_channels.size() << "channels from database";
}

QVector<ChannelInfo> ChannelManager::allChannels() const
{
    return m_channels.values().toVector();
}

ChannelInfo ChannelManager::channelInfo(int channelId) const
{
    return m_channels.value(channelId, ChannelInfo{});
}

QVector<ChannelInfo> ChannelManager::channelsByStatus(ChannelStatus status) const
{
    QVector<ChannelInfo> result;
    for (const auto& ch : m_channels) {
        if (ch.status == status) {
            result.append(ch);
        }
    }
    return result;
}

bool ChannelManager::bindProduct(int channelId, const QString& sn, const QString& pn)
{
    if (!m_channels.contains(channelId)) {
        qWarning() << "Cannot bind: channel" << channelId << "not found";
        return false;
    }

    auto& ch = m_channels[channelId];
    if (ch.status == ChannelStatus::Testing) {
        qWarning() << "Cannot bind: channel" << channelId << "is currently testing";
        return false;
    }

    ch.boundSN = sn;
    ch.boundPN = pn;
    ch.status = ChannelStatus::Bound;

    DatabaseManager::instance().localDb()->saveChannelInfo(ch);

    emit channelStatusChanged(channelId, ch.status);
    emit bindingChanged(channelId, sn, pn);

    qDebug() << "Bound product SN:" << sn << "PN:" << pn << "to channel" << channelId;
    return true;
}

bool ChannelManager::unbindProduct(int channelId)
{
    if (!m_channels.contains(channelId)) {
        qWarning() << "Cannot unbind: channel" << channelId << "not found";
        return false;
    }

    auto& ch = m_channels[channelId];

    // If channel was testing, power off first
    if (ch.status == ChannelStatus::Testing) {
        powerOff(channelId);
    }

    QString sn = ch.boundSN;
    QString pn = ch.boundPN;

    ch.boundSN.clear();
    ch.boundPN.clear();
    ch.status = ChannelStatus::Free;

    DatabaseManager::instance().localDb()->saveChannelInfo(ch);

    emit channelStatusChanged(channelId, ch.status);
    emit bindingChanged(channelId, QString(), QString());

    qDebug() << "Unbound product from channel" << channelId;
    return true;
}

bool ChannelManager::powerOn(int channelId)
{
    if (!m_channels.contains(channelId)) {
        qWarning() << "Cannot power on: channel" << channelId << "not found";
        return false;
    }

    auto& ch = m_channels[channelId];
    if (ch.status != ChannelStatus::Bound && ch.status != ChannelStatus::Testing) {
        qWarning() << "Cannot power on: channel" << channelId
                   << "is not bound or testing, status:" << static_cast<int>(ch.status);
        return false;
    }

    // Initialize power controller if needed
    if (!m_powerControllers.contains(ch.powerControllerId) || ch.powerControllerId < 0) {
        if (!initializeController(ch)) {
            qWarning() << "Failed to initialize controller for channel" << channelId;
            return false;
        }
    }

    // Turn on contactor first, then power supply
    bool success = true;

    if (ch.contactorControllerId >= 0 && m_contactorControllers.contains(ch.contactorControllerId)) {
        if (!m_contactorControllers.value(ch.contactorControllerId)->powerOn(ch.contactorChannel)) {
            qWarning() << "Failed to turn on contactor for channel" << channelId;
            success = false;
        }
    }

    if (ch.powerControllerId >= 0 && m_powerControllers.contains(ch.powerControllerId)) {
        if (!m_powerControllers.value(ch.powerControllerId)->powerOn(ch.powerChannel)) {
            qWarning() << "Failed to turn on power supply for channel" << channelId;
            success = false;
            // Turn off contactor on failure
            if (ch.contactorControllerId >= 0 && m_contactorControllers.contains(ch.contactorControllerId)) {
                m_contactorControllers.value(ch.contactorControllerId)->powerOff(ch.contactorChannel);
            }
        }
    }

    if (success) {
        ch.status = ChannelStatus::Testing;
        DatabaseManager::instance().localDb()->saveChannelInfo(ch);
        emit channelStatusChanged(channelId, ch.status);
        emit powerStateChanged(channelId, true);
        qDebug() << "Powered on channel" << channelId;
    }

    return success;
}

bool ChannelManager::powerOff(int channelId)
{
    if (!m_channels.contains(channelId)) {
        qWarning() << "Cannot power off: channel" << channelId << "not found";
        return false;
    }

    auto& ch = m_channels[channelId];

    bool success = true;

    // Turn off power supply first, then contactor
    if (ch.powerControllerId >= 0 && m_powerControllers.contains(ch.powerControllerId)) {
        if (!m_powerControllers.value(ch.powerControllerId)->powerOff(ch.powerChannel)) {
            qWarning() << "Failed to turn off power supply for channel" << channelId;
            success = false;
        }
    }

    if (ch.contactorControllerId >= 0 && m_contactorControllers.contains(ch.contactorControllerId)) {
        if (!m_contactorControllers.value(ch.contactorControllerId)->powerOff(ch.contactorChannel)) {
            qWarning() << "Failed to turn off contactor for channel" << channelId;
            success = false;
        }
    }

    if (success && ch.status == ChannelStatus::Testing) {
        ch.status = ChannelStatus::Bound;
        DatabaseManager::instance().localDb()->saveChannelInfo(ch);
        emit channelStatusChanged(channelId, ch.status);
    }

    emit powerStateChanged(channelId, false);
    qDebug() << "Powered off channel" << channelId;
    return success;
}

bool ChannelManager::setPower(int channelId, double voltage, double current)
{
    if (!m_channels.contains(channelId)) {
        qWarning() << "Cannot set power: channel" << channelId << "not found";
        return false;
    }

    auto& ch = m_channels[channelId];

    if (ch.powerControllerId >= 0 && m_powerControllers.contains(ch.powerControllerId)) {
        auto* ctrl = m_powerControllers.value(ch.powerControllerId);
        if (!ctrl->setVoltage(ch.powerChannel, voltage)) {
            qWarning() << "Failed to set voltage for channel" << channelId;
            return false;
        }
        if (!ctrl->setCurrent(ch.powerChannel, current)) {
            qWarning() << "Failed to set current for channel" << channelId;
            return false;
        }
        qDebug() << "Set power for channel" << channelId
                 << "voltage:" << voltage << "current:" << current;
        return true;
    }

    qWarning() << "No power controller available for channel" << channelId;
    return false;
}

void ChannelManager::emergencyStopAll()
{
    qDebug() << "Emergency stop: powering off all channels";

    for (auto it = m_channels.begin(); it != m_channels.end(); ++it) {
        int channelId = it.key();
        if (it.value().status == ChannelStatus::Testing) {
            powerOff(channelId);
        }
    }
}

bool ChannelManager::initializeController(ChannelInfo& info)
{
    // Initialize power controller
    if (!info.powerProtocolName.isEmpty() && info.powerControllerId >= 0) {
        if (!m_powerControllers.contains(info.powerControllerId)) {
            auto* ctrl = PluginManager::instance().createController(info.powerProtocolName);
            if (ctrl) {
                DeviceConfig config;
                config.deviceId = info.powerControllerId;
                config.protocolName = info.powerProtocolName;
                config.channelId = info.channelId;
                if (ctrl->connect(config)) {
                    m_powerControllers.insert(info.powerControllerId, ctrl);
                    qDebug() << "Initialized power controller for channel" << info.channelId;
                } else {
                    qWarning() << "Failed to connect power controller for channel" << info.channelId;
                    return false;
                }
            } else {
                qWarning() << "Failed to create power controller with protocol:" << info.powerProtocolName;
                return false;
            }
        }
    }

    // Initialize contactor controller
    if (!info.contactorProtocolName.isEmpty() && info.contactorControllerId >= 0) {
        if (!m_contactorControllers.contains(info.contactorControllerId)) {
            auto* ctrl = PluginManager::instance().createController(info.contactorProtocolName);
            if (ctrl) {
                DeviceConfig config;
                config.deviceId = info.contactorControllerId;
                config.protocolName = info.contactorProtocolName;
                config.channelId = info.channelId;
                if (ctrl->connect(config)) {
                    m_contactorControllers.insert(info.contactorControllerId, ctrl);
                    qDebug() << "Initialized contactor controller for channel" << info.channelId;
                } else {
                    qWarning() << "Failed to connect contactor controller for channel" << info.channelId;
                    return false;
                }
            } else {
                qWarning() << "Failed to create contactor controller with protocol:" << info.contactorProtocolName;
                return false;
            }
        }
    }

    return true;
}

File: src/instrument/channel_manager.py
"""Channel management and power control.

Manages the lifecycle of test channels: loading from the database,
binding/unbinding products, and controlling power supply and contactor
instruments through the plugin system.
"""

import logging

from PyQt6.QtCore import QObject, pyqtSignal

from ..core.common.types import ChannelStatus
from ..core.common.channel_info import ChannelInfo

logger = logging.getLogger(__name__)


class ChannelManager(QObject):
    """Manages test channels and their associated power/contactor controllers.

    Signals:
        channel_status_changed: Emitted when a channel's status changes.
        binding_changed: Emitted when a channel is bound or unbound.
        power_state_changed: Emitted when power is turned on or off.
    """

    channel_status_changed = pyqtSignal(int, object)   # channel_id, ChannelStatus
    binding_changed = pyqtSignal(int, str, str)          # channel_id, sn, pn
    power_state_changed = pyqtSignal(int, bool)          # channel_id, powered

    def __init__(self, parent=None):
        super().__init__(parent)
        # channel_id -> ChannelInfo
        self._channels: dict[int, ChannelInfo] = {}
        # channel_id -> IInstrumentController (power supply)
        self._power_controllers: dict[int, object] = {}
        # channel_id -> IInstrumentController (contactor)
        self._contactor_controllers: dict[int, object] = {}
        # PluginManager reference (injected)
        self._plugin_manager = None

    # ------------------------------------------------------------------
    # Dependencies
    # ------------------------------------------------------------------

    def set_plugin_manager(self, plugin_manager):
        """Set the :class:`PluginManager` used to create controllers."""
        self._plugin_manager = plugin_manager

    # ------------------------------------------------------------------
    # Loading
    # ------------------------------------------------------------------

    def load_channels(self):
        """Load all channel configurations from the database."""
        from ..storage.database_manager import DatabaseManager

        try:
            rows = DatabaseManager.instance().local_db.load_all_channels()
            for row in rows:
                status_val = row.get("status", 0)
                # Map stored integer to ChannelStatus enum
                try:
                    status = list(ChannelStatus)[status_val]
                except (IndexError, TypeError):
                    status = ChannelStatus.FREE
                ch = ChannelInfo(
                    channel_id=row["channel_id"],
                    status=status,
                    bound_sn=row.get("bound_sn", ""),
                    bound_pn=row.get("bound_pn", ""),
                    power_controller_id=row.get("power_controller_id", -1),
                    contactor_controller_id=row.get(
                        "contactor_controller_id", -1
                    ),
                    power_channel=row.get("power_channel", -1),
                    contactor_channel=row.get("contactor_channel", -1),
                    power_protocol_name=row.get("power_protocol_name", ""),
                    contactor_protocol_name=row.get(
                        "contactor_protocol_name", ""
                    ),
                )
                self._channels[ch.channel_id] = ch
            logger.info("Loaded %d channels", len(self._channels))
        except Exception as e:
            logger.error("Failed to load channels: %s", e)

    # ------------------------------------------------------------------
    # Query
    # ------------------------------------------------------------------

    def all_channels(self) -> list[ChannelInfo]:
        """Return a copy of all channel info objects."""
        return list(self._channels.values())

    def channel_info(self, channel_id: int) -> ChannelInfo | None:
        """Return the :class:`ChannelInfo` for *channel_id*, or None."""
        return self._channels.get(channel_id)

    # ------------------------------------------------------------------
    # Binding
    # ------------------------------------------------------------------

    def bind_product(self, channel_id: int, sn: str, pn: str) -> bool:
        """Bind a product (serial number / part number) to a channel."""
        from ..storage.database_manager import DatabaseManager

        ch = self._channels.get(channel_id)
        if not ch:
            return False
        ch.bound_sn = sn
        ch.bound_pn = pn
        ch.status = ChannelStatus.BOUND
        try:
            self._persist_channel(ch)
        except Exception as e:
            logger.error("Failed to save channel: %s", e)
        self.channel_status_changed.emit(channel_id, ch.status)
        self.binding_changed.emit(channel_id, sn, pn)
        return True

    def unbind_product(self, channel_id: int) -> bool:
        """Unbind a product from a channel (powers off if testing)."""
        from ..storage.database_manager import DatabaseManager

        ch = self._channels.get(channel_id)
        if not ch:
            return False
        if ch.status == ChannelStatus.TESTING:
            self.power_off(channel_id)
        ch.bound_sn = ""
        ch.bound_pn = ""
        ch.status = ChannelStatus.FREE
        try:
            self._persist_channel(ch)
        except Exception as e:
            logger.error("Failed to save channel: %s", e)
        self.channel_status_changed.emit(channel_id, ch.status)
        self.binding_changed.emit(channel_id, "", "")
        return True

    # ------------------------------------------------------------------
    # Power control
    # ------------------------------------------------------------------

    def power_on(self, channel_id: int) -> bool:
        """Turn on power for a channel (contactor first, then supply)."""
        ch = self._channels.get(channel_id)
        if not ch:
            return False
        try:
            # Turn on contactor first
            if ch.contactor_protocol_name:
                ctrl = self._contactor_controllers.get(channel_id)
                if not ctrl:
                    ctrl = self._init_controller(ch, "contactor")
                    if ctrl:
                        self._contactor_controllers[channel_id] = ctrl
                if ctrl:
                    ctrl.power_on(ch.contactor_channel)
            # Then power supply
            if ch.power_protocol_name:
                ctrl = self._power_controllers.get(channel_id)
                if not ctrl:
                    ctrl = self._init_controller(ch, "power")
                    if ctrl:
                        self._power_controllers[channel_id] = ctrl
                if ctrl:
                    ctrl.power_on(ch.power_channel)
            ch.status = ChannelStatus.TESTING
            self._persist_channel(ch)
            self.channel_status_changed.emit(channel_id, ch.status)
            self.power_state_changed.emit(channel_id, True)
            return True
        except Exception as e:
            logger.error("Power on failed for channel %d: %s", channel_id, e)
            return False

    def power_off(self, channel_id: int) -> bool:
        """Turn off power for a channel (supply first, then contactor)."""
        ch = self._channels.get(channel_id)
        if not ch:
            return False
        try:
            # Turn off power supply first
            ctrl = self._power_controllers.get(channel_id)
            if ctrl:
                try:
                    ctrl.power_off(ch.power_channel)
                except Exception:
                    pass
            # Then contactor
            ctrl = self._contactor_controllers.get(channel_id)
            if ctrl:
                try:
                    ctrl.power_off(ch.contactor_channel)
                except Exception:
                    pass
            ch.status = (
                ChannelStatus.BOUND
                if ch.bound_sn
                else ChannelStatus.FREE
            )
            self._persist_channel(ch)
            self.channel_status_changed.emit(channel_id, ch.status)
            self.power_state_changed.emit(channel_id, False)
            return True
        except Exception as e:
            logger.error("Power off failed for channel %d: %s", channel_id, e)
            return False

    def emergency_stop_all(self):
        """Power off all channels that are currently testing."""
        for ch_id, ch in list(self._channels.items()):
            if ch.status == ChannelStatus.TESTING:
                self.power_off(ch_id)
        logger.warning("Emergency stop all channels")

    # ------------------------------------------------------------------
    # Controller access (for RecipeExecutor)
    # ------------------------------------------------------------------

    def get_power_controller(self, channel_id: int):
        """Return the power supply controller for *channel_id*, or None."""
        return self._power_controllers.get(channel_id)

    def get_contactor_controller(self, channel_id: int):
        """Return the contactor controller for *channel_id*, or None."""
        return self._contactor_controllers.get(channel_id)

    def execute_relay_action(self, channel_id: int, relay_id: int, action: str) -> bool:
        """Execute a single relay action on the channel's contactor controller.

        Args:
            channel_id: Channel identifier.
            relay_id: Coil number on the contactor controller.
            action: ``"CLOSE"`` or ``"OPEN"``.
        """
        ch = self._channels.get(channel_id)
        if not ch:
            return False
        ctrl = self._contactor_controllers.get(channel_id)
        if not ctrl:
            ctrl = self._init_controller(ch, "contactor")
            if ctrl:
                self._contactor_controllers[channel_id] = ctrl
        if not ctrl:
            logger.warning(
                "No contactor controller for channel %d, relay action skipped",
                channel_id,
            )
            return False
        try:
            if action.upper() == "CLOSE":
                ctrl.power_on(relay_id)
            else:
                ctrl.power_off(relay_id)
            return True
        except Exception as e:
            logger.error("Relay action failed ch%d relay%d: %s", channel_id, relay_id, e)
            return False

    def set_channel_output(self, channel_id: int, voltage: float, current: float) -> bool:
        """Set voltage and current on the channel's power supply controller."""
        ch = self._channels.get(channel_id)
        if not ch:
            return False
        ctrl = self._power_controllers.get(channel_id)
        if not ctrl:
            ctrl = self._init_controller(ch, "power")
            if ctrl:
                self._power_controllers[channel_id] = ctrl
        if not ctrl:
            logger.warning("No power controller for channel %d", channel_id)
            return False
        try:
            ctrl.set_voltage(ch.power_channel, voltage)
            ctrl.set_current(ch.power_channel, current)
            return True
        except Exception as e:
            logger.error("Set output failed ch%d: %s", channel_id, e)
            return False

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _init_controller(self, ch: ChannelInfo, controller_type: str):
        """Create and connect a power or contactor controller for *ch*."""
        if not self._plugin_manager:
            logger.error("PluginManager not set; cannot create controller")
            return None
        from ..core.common.types import DeviceConfig

        protocol_name = (
            ch.power_protocol_name
            if controller_type == "power"
            else ch.contactor_protocol_name
        )
        if not protocol_name:
            return None

        config = DeviceConfig(
            device_id=(
                ch.power_controller_id
                if controller_type == "power"
                else ch.contactor_controller_id
            ),
            protocol_name=protocol_name,
        )
        controller = self._plugin_manager.create_controller(protocol_name)
        if controller:
            try:
                controller.connect(config)
            except Exception as e:
                logger.error(
                    "Failed to connect %s controller: %s",
                    controller_type,
                    e,
                )
                return None
        return controller

    def _persist_channel(self, ch: ChannelInfo):
        """Write channel info to the database as a dict."""
        from ..storage.database_manager import DatabaseManager

        DatabaseManager.instance().local_db.save_channel_info({
            "channel_id": ch.channel_id,
            "status": ch.status.value if isinstance(ch.status, ChannelStatus) else ch.status,
            "bound_sn": ch.bound_sn,
            "bound_pn": ch.bound_pn,
            "power_controller_id": ch.power_controller_id,
            "contactor_controller_id": ch.contactor_controller_id,
            "power_channel": ch.power_channel,
            "contactor_channel": ch.contactor_channel,
            "power_protocol_name": ch.power_protocol_name,
            "contactor_protocol_name": ch.contactor_protocol_name,
        })
